File: ios-app/RunningTrainerNative/RunningTrainer/Domain/Services/ProgressStatsCalculator.swift
import Foundation

final class ProgressStatsCalculator {
    private let today: () -> Date

    init(today: @escaping () -> Date = { Date() }) {
        self.today = today
    }

    func compute(plan: TrainingPlan) -> ProgressStats {
        let cal = Calendar.current
        let now = today()
        let daysSinceStart = cal.dateComponents([.day], from: plan.startDate.startOfDay, to: now.startOfDay).day ?? 0
        let currentWeekIndex = min(max(daysSinceStart / 7, 0), plan.totalWeeks - 1)

        var totalNonRest = 0
        var completed = 0
        var totalPlanned = 0.0
        var totalLogged = 0.0
        var weeklyProgress: [WeekProgress] = []
        var rpePoints: [RpeDataPoint] = []
        var feelingMap: [WorkoutFeeling: Int] = [:]
        var workoutTypeMap: [WorkoutType: Int] = [:]

        for (weekIndex, week) in plan.weeks.enumerated() {
            let hasStarted = weekIndex <= currentWeekIndex
            var weekPlanned = 0.0
            var weekLogged = 0.0
            var weekNonRest = 0
            var weekCompleted = 0

            for workout in week.workouts {
                if workout.type == .rest { continue }

                weekNonRest += 1
                totalNonRest += 1
                weekPlanned += workout.distanceKm ?? 0.0
                totalPlanned += workout.distanceKm ?? 0.0

                if hasStarted {
                    workoutTypeMap[workout.type, default: 0] += 1
                }

                if workout.isCompleted {
                    weekCompleted += 1
                    completed += 1
                    let loggedKm = workout.actualDistanceKm ?? workout.distanceKm ?? 0.0
                    weekLogged += loggedKm
                    totalLogged += loggedKm
                    if let rpe = workout.rpe, let completedAt = workout.completedAt {
                        rpePoints.append(RpeDataPoint(date: completedAt, rpe: rpe, type: workout.type))
                    }
                    if let feeling = workout.feeling {
                        feelingMap[feeling, default: 0] += 1
                    }
                }
            }

            if hasStarted {
                weeklyProgress.append(WeekProgress(
                    weekNumber: weekIndex + 1,
                    plannedKm: weekPlanned,
                    loggedKm: weekLogged,
                    totalWorkouts: weekNonRest,
                    completedWorkouts: weekCompleted,
                    hasStarted: true
                ))
            }
        }

        let recentRpe = Array(rpePoints.sorted { $0.date < $1.date }.suffix(12))

        let allCompleted = plan.weeks
            .flatMap { $0.workouts }
            .filter { $0.isCompleted && $0.type != .rest }
            .sorted { ($0.completedAt ?? .distantPast) < ($1.completedAt ?? .distantPast) }

        let pacePoints = allCompleted
            .filter { ($0.actualDistanceKm ?? 0) > 0 && ($0.actualDurationMinutes ?? 0) > 0 }
            .suffix(12)
            .compactMap { w -> PaceDataPoint? in
                guard let completedAt = w.completedAt else { return nil }
                return PaceDataPoint(
                    paceMinPerKm: Double(w.actualDurationMinutes!) / w.actualDistanceKm!,
                    type: w.type,
                    date: completedAt
                )
            }

        let streak = computeStreak(plan: plan, today: now)

        return ProgressStats(
            totalNonRestWorkouts: totalNonRest,
            completedWorkouts: completed,
            totalPlannedKm: totalPlanned,
            totalLoggedKm: totalLogged,
            currentStreak: streak,
            weeklyProgress: weeklyProgress,
            rpeDataPoints: recentRpe,
            feelingCounts: feelingMap,
            paceDataPoints: pacePoints,
            workoutTypeCounts: WorkoutType.allCases
                .filter { $0 != .rest }
                .map { WorkoutTypeCount(type: $0, count: workoutTypeMap[$0] ?? 0) }
                .filter { $0.count > 0 },
            recentCompletedWorkouts: Array(allCompleted.suffix(8).reversed())
        )
    }

    private func computeStreak(plan: TrainingPlan, today: Date) -> Int {
        let cal = Calendar.current
        var streak = 0
        for offset in stride(from: 0, through: -365, by: -1) {
            guard let date = cal.date(byAdding: .day, value: offset, to: today) else { break }
            let daysSince = cal.dateComponents([.day], from: plan.startDate.startOfDay, to: date.startOfDay).day ?? -1
            if daysSince < 0 { break }
            let weekIndex = daysSince / 7
            if weekIndex >= plan.weeks.count { continue }
            let dow1Mon = ((cal.component(.weekday, from: date) + 5) % 7) + 1
            guard let workout = plan.weeks[weekIndex].workouts.first(where: { $0.dayOfWeek == dow1Mon }) else { continue }
            if workout.type == .rest { continue }
            if !workout.isCompleted { break }
            streak += 1
        }
        return streak
    }
}

private extension Date {
    var startOfDay: Date { Calendar.current.startOfDay(for: self) }
}

File: web-app/src/domain/progressStatsCalculator.ts
import type {
  ProgressStats,
  TrainingPlan,
  WeekProgress,
  WorkoutFeeling,
  WorkoutType,
} from './models'

function todayStr(): string {
  return new Date().toISOString().slice(0, 10)
}

function daysBetween(from: string, to: string): number {
  return Math.round((new Date(to).getTime() - new Date(from).getTime()) / 86400000)
}

function addDays(dateStr: string, n: number): string {
  const d = new Date(dateStr)
  d.setDate(d.getDate() + n)
  return d.toISOString().slice(0, 10)
}

function isoDow(dateStr: string): number {
  const d = new Date(dateStr).getDay()
  return d === 0 ? 7 : d
}

export function computeProgressStats(plan: TrainingPlan): ProgressStats {
  const today = todayStr()
  const daysSinceStart = daysBetween(plan.startDate, today)
  const currentWeekIndex = Math.max(0, Math.min(Math.floor(daysSinceStart / 7), plan.totalWeeks - 1))

  let totalNonRest = 0
  let completed = 0
  let totalPlannedKm = 0
  let totalLoggedKm = 0
  const weeklyProgress: WeekProgress[] = []
  const rpePoints: ProgressStats['rpeDataPoints'] = []
  const feelingCounts: Record<WorkoutFeeling, number> = {
    great: 0, good: 0, ok: 0, tired: 0, injured: 0,
  }
  const workoutTypeMap: Partial<Record<WorkoutType, number>> = {}

  plan.weeks.forEach((week, weekIndex) => {
    const hasStarted = weekIndex <= currentWeekIndex
    let weekPlanned = 0
    let weekLogged = 0
    let weekNonRest = 0
    let weekCompleted = 0

    week.workouts.forEach((workout) => {
      if (workout.type === 'rest') return

      weekNonRest++
      totalNonRest++
      weekPlanned += workout.distanceKm ?? 0
      totalPlannedKm += workout.distanceKm ?? 0

      if (hasStarted) {
        workoutTypeMap[workout.type] = (workoutTypeMap[workout.type] ?? 0) + 1
      }

      if (workout.isCompleted) {
        weekCompleted++
        completed++
        const loggedKm = workout.actualDistanceKm ?? workout.distanceKm ?? 0
        weekLogged += loggedKm
        totalLoggedKm += loggedKm
        if (workout.rpe != null && workout.completedAt != null) {
          rpePoints.push({
            date: workout.completedAt.slice(0, 10),
            rpe: workout.rpe,
            type: workout.type,
          })
        }
        if (workout.feeling != null) {
          feelingCounts[workout.feeling] = (feelingCounts[workout.feeling] ?? 0) + 1
        }
      }
    })

    if (hasStarted) {
      weeklyProgress.push({
        weekNumber: weekIndex + 1,
        plannedKm: weekPlanned,
        loggedKm: weekLogged,
        totalWorkouts: weekNonRest,
        completedWorkouts: weekCompleted,
        hasStarted: true,
      })
    }
  })

  const allCompleted = plan.weeks
    .flatMap((w) => w.workouts)
    .filter((w) => w.isCompleted && w.type !== 'rest')
    .sort((a, b) => (a.completedAt ?? '').localeCompare(b.completedAt ?? ''))

  const paceDataPoints = allCompleted
    .filter((w) => (w.actualDistanceKm ?? 0) > 0 && (w.actualDurationMinutes ?? 0) > 0)
    .slice(-12)
    .map((w) => ({
      paceMinPerKm: w.actualDurationMinutes! / w.actualDistanceKm!,
      type: w.type,
      date: (w.completedAt ?? '').slice(0, 10),
    }))

  const streak = computeStreak(plan, today)

  const workoutTypeCounts = (Object.entries(workoutTypeMap) as [WorkoutType, number][])
    .map(([type, count]) => ({ type, count }))
    .filter((x) => x.count > 0)

  return {
    totalNonRestWorkouts: totalNonRest,
    completedWorkouts: completed,
    totalPlannedKm,
    totalLoggedKm,
    currentStreak: streak,
    weeklyProgress,
    rpeDataPoints: rpePoints.slice(-12),
    feelingCounts,
    paceDataPoints,
    workoutTypeCounts,
    recentCompletedWorkouts: [...allCompleted].slice(-8).reverse(),
  }
}

function computeStreak(plan: TrainingPlan, today: string): number {
  let streak = 0
  for (let offset = 0; offset >= -365; offset--) {
    const date = addDays(today, offset)
    const ds = daysBetween(plan.startDate, date)
    if (ds < 0) break
    const wi = Math.floor(ds / 7)
    if (wi >= plan.weeks.length) continue
    const dow = isoDow(date)
    const workout = plan.weeks[wi].workouts.find((w) => w.dayOfWeek === dow)
    if (!workout || workout.type === 'rest') continue
    if (!workout.isCompleted) break
    streak++
  }
  return streak
}
